File: desktop/src/types/mcp.ts
/**
 * MCP server types matching tools/mcp_tool.py.
 * @source tools/mcp_tool.py
 */

/** MCP server transport type. */
export type McpTransport = 'stdio' | 'http' | 'streamable_http' | 'sse';

/** MCP server authentication type. */
export type McpAuthType = 'oauth' | 'bearer' | 'api_key' | null;

/** Desktop-only MCP server metadata. */
export interface McpServerDesktop {
  pinned: boolean;
  note?: string | null;
  display_order?: number | null;
  last_selected_at?: string | null;
  updated_at?: string | null;
}

/** Desktop-only MCP metadata patch. */
export interface McpServerDesktopPatch {
  pinned?: boolean;
  note?: string | null;
  display_order?: number | null;
  last_selected_at?: string | null;
}

/** MCP server configuration. */
export interface McpServer {
  name: string;
  command?: string;
  args?: string[];
  env?: Record<string, string>;
  url?: string;
  headers?: Record<string, string>;
  timeout?: number;
  connect_timeout?: number;
  auth?: McpAuthType;
  oauth?: McpOAuthConfig;
  sampling?: McpSamplingConfig;
  transport?: McpTransport;
  enabled?: boolean;
  valid?: boolean;
  error?: string | null;
  status?: Record<string, unknown> | null;
  desktop?: McpServerDesktop;
}

/** MCP OAuth configuration. */
export interface McpOAuthConfig {
  client_id?: string;
  client_secret?: string;
  auth_url?: string;
  token_url?: string;
  scopes?: string[];
}

/** MCP sampling configuration for server-initiated LLM requests. */
export interface McpSamplingConfig {
  enabled?: boolean;
  model?: string;
  max_tokens_cap?: number;
  timeout?: number;
  max_rpm?: number;
  allowed_models?: string[];
  max_tool_rounds?: number;
  log_level?: 'debug' | 'info' | 'warning';
}

/** MCP tool as returned by list_tools. */
export interface McpTool {
  name: string;
  description?: string;
  inputSchema?: McpInputSchema;
}

/** MCP input schema for a tool. */
export interface McpInputSchema {
  type?: string;
  properties?: Record<string, McpSchemaProperty>;
  required?: string[];
}

/** MCP schema property. */
export interface McpSchemaProperty {
  type?: string;
  description?: string;
  default?: unknown;
  enum?: unknown[];
}

/** MCP server connection state. */
export type McpRuntimeStatus =
  | 'connected'
  | 'connecting'
  | 'configured'
  | 'failed'
  | 'disabled'
  | string;

export interface McpConnectionStatus {
  name: string;
  connected: boolean;
  transport: McpTransport;
  tools: number;
  error?: string;
  disabled?: boolean;
  status?: McpRuntimeStatus;
}

File: desktop/sidecar/daemon/schemas/mcp.py
from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, model_validator


McpTransport = Literal["stdio", "http", "streamable_http", "sse"]


class McpServerDesktop(BaseModel):
    pinned: bool = False
    note: Optional[str] = None
    display_order: Optional[int] = None
    last_selected_at: Optional[str] = None
    updated_at: Optional[str] = None


class PatchMcpServerDesktopRequest(BaseModel):
    pinned: Optional[bool] = None
    note: Optional[str] = None
    display_order: Optional[int] = None
    last_selected_at: Optional[str] = None


class McpServerConfig(BaseModel):
    name: str
    command: Optional[str] = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)
    url: Optional[str] = None
    headers: dict[str, str] = Field(default_factory=dict)
    timeout: Optional[int] = None
    connect_timeout: Optional[int] = None
    auth: Optional[str] = None
    oauth: Optional[dict[str, Any]] = None
    sampling: Optional[dict[str, Any]] = None
    transport: McpTransport = "stdio"
    enabled: bool = True


class McpServerCreate(McpServerConfig):
    name: str = Field(min_length=1, pattern=r"^[A-Za-z0-9_.-]+$")

    @model_validator(mode="after")
    def validate_transport_shape(self) -> "McpServerCreate":
        if self.transport == "stdio" and not self.command:
            raise ValueError("stdio MCP servers require command")
        if self.transport != "stdio" and not self.url:
            raise ValueError("HTTP MCP servers require url")
        return self


class McpServer(McpServerConfig):
    valid: bool = True
    error: Optional[str] = None
    status: Optional[dict[str, Any]] = None
    desktop: McpServerDesktop = Field(default_factory=McpServerDesktop)


class McpServerList(BaseModel):
    items: list[McpServer]
    generated_at: str


class McpReloadResponse(McpServerList):
    ok: bool = True
    refreshed_agents: int = 0


class McpTool(BaseModel):
    name: str
    description: Optional[str] = None
    inputSchema: Optional[dict[str, Any]] = None


class McpToolList(BaseModel):
    items: list[McpTool]
    status: Optional[dict[str, Any]] = None
